docs/multi_credential_assignment.md

```diff
@@ -0,0 +1,163 @@
+Multi-Credential Assignment
+===========================
+
+awx has added support for assigning zero or more credentials to
+a JobTemplate via a singular, unified interface.
+
+Background
+----------
+
+Prior to awx (Tower 3.2), Job Templates had a certain set of requirements
+surrounding their relation to Credentials:
+
+* All Job Templates (and Jobs) were required to have exactly *one* Machine/SSH
+  or Vault credential (or one of both).
+* All Job Templates (and Jobs) could have zero or more "extra" Credentials.
+* These extra Credentials represented "Cloud" and "Network" credentials that
+* could be used to provide authentication to external services via environment
+* variables (e.g., AWS_ACCESS_KEY_ID).
+
+This model required a variety of disjoint interfaces for specifying Credentials
+on a JobTemplate.  For example, to modify assignment of Machine/SSH and Vault
+credentials, you would change the Credential key itself:
+
+`PATCH /api/v2/job_templates/N/ {'credential': X, 'vault_credential': Y}`
+
+Modifying `extra_credentials` was accomplished on a separate API endpoint
+via association/disassociation actions:
+
+```
+POST /api/v2/job_templates/N/extra_credentials {'associate': true, 'id': Z}
+POST /api/v2/job_templates/N/extra_credentials {'disassociate': true, 'id': Z}
+```
+
+This model lacked the ability associate multiple Vault credentials with
+a playbook run, a use case supported by Ansible core from Ansible 2.4 onwards.
+
+This model also was a stumbling block for certain playbook execution workflows.
+For example, some users wanted to run playbooks with `connection:local` that
+only interacted with some cloud service via a cloud Credential.  In this
+scenario, users often generated a "dummy" Machine/SSH Credential to attach to
+the Job Template simply to satisfy the requirement on the model.
+
+Important Changes
+-----------------
+
+JobTemplates now have a single interface for Credential assignment:
+
+`GET /api/v2/job_templates/N/credentials/`
+
+Users can associate and disassociate credentials using `POST` requests to this
+interface, similar to the behavior in the now-deprecated `extra_credentials`
+endpoint:
+
+```
+POST /api/v2/job_templates/N/credentials/ {'associate': true, 'id': X'}
+POST /api/v2/job_templates/N/credentials/ {'disassociate': true, 'id': Y'}
+```
+
+Under this model, a JobTemplate is considered valid even when it has _zero_
+Credentials assigned to it.
+
+Launch Time Considerations
+--------------------------
+
+Prior to this change, JobTemplates had a configurable attribute,
+`ask_credential_on_launch`.  This value was used at launch time to determine
+which missing credential values were necessary for launch - this was primarily
+used as a mechanism for users to specify a Machine/SSH credential to satisfy 
```

Comment:
It let users specify either a Machine credential or a vault credential.

This needs to be said, but I find this wording a little confusing. Previously, a non-null credential requirement (with caveats about vault) was enforced, and that changed here.

Reply:
👍 I'll update this to mention Vault.